include/zephyr/drivers/retained_mem.h

```diff
@@ -0,0 +1,198 @@
+/**
+ * @typedef	retained_mem_read_api
+ * @brief	Callback API to read from retained memory area.
+ * See retained_mem_read() for argument description.
+ */
+typedef int (*retained_mem_read_api)(const struct device *dev, off_t offset, uint8_t *buffer,
+				     size_t size);
+
+/**
+ * @typedef	retained_mem_write_api
+ * @brief	Callback API to write to retained memory area.
+ * See retained_mem_write() for argument description.
+ */
+typedef int (*retained_mem_write_api)(const struct device *dev, off_t offset,
+				      const uint8_t *buffer, size_t size);
+
+/**
+ * @typedef	retained_mem_clear_api
+ * @brief	Callback API to clear retained memory area (reset all data to 0x00).
+ * See retained_mem_clear() for argument description.
+ */
+typedef int (*retained_mem_clear_api)(const struct device *dev);
+
+/**
+ * @brief Retained memory driver API
+ * API which can be used by a device to store data in a retained memory area. Retained memory is
+ * memory that is retained while the device is powered but is lost when power to the device is
+ * lost (note that low power modes in some devices may clear the data also). This may be in a
+ * non-initialised RAM region, or in specific registers, but is not reset when a different
+ * application begins execution or the device is rebooted (without power loss). It must support
+ * byte-level reading and writing without a need to erase data before writing.
+ *
+ * Note that drivers must implement all functions, none of the functions are optional.
+ */
+struct retained_mem_driver_api {
+	retained_mem_size_api size;
+	retained_mem_read_api read;
+	retained_mem_write_api write;
+	retained_mem_clear_api clear;
+};
+
+/**
+ * @brief		Returns the size of the retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ *
+ * @retval		Positive value indicating size in bytes on success, else negative errno
+ *			code.
+ */
+__syscall ssize_t retained_mem_size(const struct device *dev);
+
+static inline ssize_t z_impl_retained_mem_size(const struct device *dev)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+
+	return api->size(dev);
+}
+
+/**
+ * @brief		Reads data from the Retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ * @param offset	Offset to read data from.
+ * @param buffer	Buffer to store read data in.
+ * @param size		Size of data to read.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int retained_mem_read(const struct device *dev, off_t offset, uint8_t *buffer,
+				size_t size);
+
+static inline int z_impl_retained_mem_read(const struct device *dev, off_t offset,
+					   uint8_t *buffer, size_t size)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+	size_t end;
+	bool overflow;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return 0;
+	} else if (offset < 0) {
+		return -EINVAL;
+	}
+
+	overflow = size_add_overflow((uint32_t)offset, size, &end);
```

Comment:
shouldn't this be a cast to `size_t` instead of `uint32_t` ?

Reply:
And can't this whole construct be simplified to:
```
if ((api->size(dev) < len) || ((api->size(dev) - size) < (size_t)offset)) {
        return -EINVAL;
}

return api->read(dev, offset, buffer, size);
```
If `api->read()` handles the 0 length gracefully even the check for for zero length could be removed. IMO the check for zero length and exiting early is an unnecessary construct as the early exiting should not occur and even if it does there is no reason to exit early.

Reply:
The current method works

Reply:
Sure it does, but why use 10 lines of code when it can be done in 3 ?

Reply:
Readability, the same reason you don't compress a 100 line function into a single line unless you're making a minified version of a library whereby you want absolutely no debugging ability.

Reply:
Regarding readability this is certainly not an example of what I would call readability. In order to understand it it is required to grep in header files to find out what `size_add_overflow` is and what it's arguments are. Finally only to be left with the question "Why has the author made this so difficult ? It is only needed to check if the box is large enough and that the starting position is not to large."